misc/multiaddr/src/protocol.rs

```diff
@@ -1,390 +1,171 @@
-            Protocol::DCCP => ProtocolArgSize::Fixed { bytes: 2 },
-            Protocol::IP6 => ProtocolArgSize::Fixed { bytes: 16 },
-            Protocol::DNS4 => ProtocolArgSize::Variable,
-            Protocol::DNS6 => ProtocolArgSize::Variable,
-            Protocol::SCTP => ProtocolArgSize::Fixed { bytes: 2 },
-            Protocol::UDT => ProtocolArgSize::Fixed { bytes: 0 },
-            Protocol::UTP => ProtocolArgSize::Fixed { bytes: 0 },
-            Protocol::UNIX => ProtocolArgSize::Variable,
-            Protocol::P2P => ProtocolArgSize::Variable,
-            Protocol::HTTP => ProtocolArgSize::Fixed { bytes: 0 },
-            Protocol::HTTPS => ProtocolArgSize::Fixed { bytes: 0 },
-            Protocol::ONION => ProtocolArgSize::Fixed { bytes: 10 },
-            Protocol::QUIC => ProtocolArgSize::Fixed { bytes: 0 },
-            Protocol::WS => ProtocolArgSize::Fixed { bytes: 0 },
-            Protocol::WSS => ProtocolArgSize::Fixed { bytes: 0 },
-            Protocol::Libp2pWebsocketStar => ProtocolArgSize::Fixed { bytes: 0 },
-            Protocol::Libp2pWebrtcStar => ProtocolArgSize::Fixed { bytes: 0 },
-            Protocol::Libp2pWebrtcDirect => ProtocolArgSize::Fixed { bytes: 0 },
-            Protocol::P2pCircuit => ProtocolArgSize::Fixed { bytes: 0 },
-            Protocol::Memory => ProtocolArgSize::Fixed { bytes: 0 },
-        }
-    }
-}
-
-#[derive(Debug, Copy, Clone, PartialEq, Eq)]
-pub enum ProtocolArgSize {
-    /// The size of the argument is of fixed length. The length can be 0, in which case there is no
-    /// argument.
-    Fixed { bytes: usize },
-    /// The size of the argument is of variable length.
-    Variable,
+/// `Protocol` describes all possible protocols.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub enum Protocol<'a> {
+    Dccp(u16),
+    Dns4(Cow<'a, str>),
+    Dns6(Cow<'a, str>),
+    Http,
+    Https,
+    Ip4(Ipv4Addr),
+    Ip6(Ipv6Addr),
+    P2pWebRtcDirect,
+    P2pWebRtcStar,
+    P2pWebSocketStar,
+    Memory,
+    Onion(Cow<'a, [u8]>),
+    P2p(Multihash),
+    P2pCircuit,
+    Quic,
+    Sctp(u16),
+    Tcp(u16),
+    Udp(u16),
+    Udt,
+    Unix(Cow<'a, str>),
+    Utp,
+    Ws,
+    Wss
 }
 
-impl Protocol {
-    /// Convert an array slice to the string representation.
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// use std::net::Ipv4Addr;
-    /// use multiaddr::AddrComponent;
-    /// use multiaddr::Protocol;
-    ///
-    /// let proto = Protocol::IP4;
-    /// assert_eq!(proto.parse_data("127.0.0.1").unwrap(),
-    ///            AddrComponent::IP4(Ipv4Addr::new(127, 0, 0, 1)));
-    /// ```
-    ///
-    pub fn parse_data<'a>(&self, a: &'a str) -> Result<AddrComponent<'a>> {
-        match *self {
-            Protocol::IP4 => {
-                let addr = Ipv4Addr::from_str(a)?;
-                Ok(AddrComponent::IP4(addr))
+impl<'a> Protocol<'a> {
+    pub fn from_str_parts<I>(mut iter: I) -> Result<Self>
```

Comment:
I would expect this method to produce an error if the iterator yields more than two elements (or one, depending on the protocol). I don't know which way is correct, but that kind of behavior should be documented.

Reply:
[This method is used](https://github.com/libp2p/rust-libp2p/pull/498/files#diff-75dfd2141539fcdccd76db7a161a7b9dL317) to parse a sequence of protocols and the iterator should be able to continue with the next protocol. Therefore `from_str_parts` can not exhaust the iterator. If this is deemed problematic we could mark `from_str_parts` as `pub(crate)` and if we want to be able to parse a `Protocol` from a`&str` alone we could add a method which checks that the iterator has been exhausted:

```rust
    pub fn from_str(s: &'a str) -> Result<Self> {
        let mut parts = s.split('/');
        let p = Self::from_str_parts(&mut parts)?;
        if parts.next().is_some() {
            Err(Error::InvalidProtocolString)
        } else {
            Ok(p)
        }
    }
```